explorer/scan/local_scanner.go

```diff
@@ -184,295 +170,194 @@ func preprocessQueryPackFilters(filters []string) []string {
-	if err != nil {
-		return nil, false, errors.New("failed to resolve inventory for connection")
-	}
-	assetList := im.GetAssets()
-
-	var assets []*assetWithRuntime
-	// note: asset candidate runtimes are the runtime that discovered them
-	var assetCandidates []*assetWithRuntime
-	var assetErrors []*assetWithError
-
-	// we connect and perform discovery for each asset in the job inventory
-	for i := range assetList {
-		asset := assetList[i]
-		resolvedAsset, err := im.ResolveAsset(asset)
-		if err != nil {
-			return nil, false, err
+func CreateProgressBar(discoveredAssets *DiscoveredAssets, disableProgressBar bool) (progress.MultiProgress, error) {
+	var multiprogress progress.MultiProgress
+	if isatty.IsTerminal(os.Stdout.Fd()) && !disableProgressBar && !strings.EqualFold(logger.GetLevel(), "debug") && !strings.EqualFold(logger.GetLevel(), "trace") {
+		progressBarElements := map[string]string{}
+		orderedKeys := []string{}
+		for i := range discoveredAssets.Assets {
+			asset := discoveredAssets.Assets[i].Asset
+			// this shouldn't happen, but might
+			// it normally indicates a bug in the provider
+			if presentAsset, present := progressBarElements[asset.PlatformIds[0]]; present {
+				return nil, fmt.Errorf("asset %s and %s have the same platform id %s", presentAsset, asset.Name, asset.PlatformIds[0])
+			}
+			progressBarElements[asset.PlatformIds[0]] = asset.Name
+			orderedKeys = append(orderedKeys, asset.PlatformIds[0])
 		}
-
-		runtime, err := providers.Coordinator.RuntimeFor(asset, providers.DefaultRuntime())
+		var err error
+		multiprogress, err = progress.NewMultiProgressBars(progressBarElements, orderedKeys, progress.WithScore())
 		if err != nil {
-			log.Error().Err(err).Str("asset", asset.Name).Msg("unable to create runtime for asset")
-			assetErrors = append(assetErrors, &assetWithError{
-				asset: resolvedAsset,
-				err:   err,
-			})
-			continue
+			return nil, multierr.Wrap(err, "failed to create progress bars")
 		}
-		runtime.SetRecording(s.recording)
-
-		if err := runtime.Connect(&plugin.ConnectReq{
-			Features: cnquery.GetFeatures(ctx),
-			Asset:    resolvedAsset,
-			Upstream: upstream,
-		}); err != nil {
-			log.Error().Err(err).Msg("unable to connect to asset")
-			assetErrors = append(assetErrors, &assetWithError{
-				asset: resolvedAsset,
-				err:   err,
-			})
-			continue
-		}
-		asset = runtime.Provider.Connection.Asset // to ensure we get all the information the connect call gave us
+	} else {
+		// TODO: adjust naming
+		multiprogress = progress.NoopMultiProgressBars{}
+	}
+	return multiprogress, nil
+}
 
-		// for all discovered assets, we apply mondoo-specific labels and annotations that come from the root asset
-		for _, a := range runtime.Provider.Connection.GetInventory().GetSpec().GetAssets() {
-			a.AddMondooLabels(asset)
-			a.AddAnnotations(asset.GetAnnotations())
+func Batch[T any](list []T, batchSize int) [][]T {
```

Comment:
might be a good one to unit test

Reply:
yup, will do that. That one used to be in cnspec. It's now moved here. I will add a test for it